File: Linked Lists/Q4. Palindrome Linked List/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
 import java.util.Stack;
class Solution {
    public ListNode rev(ListNode head){
        ListNode a=null, b=head, c=null;
        while(b!=null){
            c=b.next;
            b.next=a;
            a=b;
            b=c;
        }
        return a;
    }
    public boolean isPalindrome(ListNode head) {
        if (head == null || head.next == null) return true;
        ListNode s = head, f=head;
        while(f!=null && f.next!=null){
            s=s.next;
            f=f.next.next;
        }
        s=rev(s);
        f=head;
        while(s!=null){
            if(f.val!=s.val){
               return false;
            }
            else{
                f=f.next;
                s=s.next;
            }
        }
        return true;
    }
}

File: Linked Lists/Q4. Palindrome Linked List/solution.py
class Solution(object):
    def isPalindrome(self, head):
        if not head or not head.next:
            return True

        # find middle of linked list
        slow = head
        fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # reverse second half
        prev = None
        while slow:
            nxt = slow.next
            slow.next = prev
            prev = slow
            slow = nxt

        # compare both halves
        left = head
        right = prev
        while right:
            if left.val != right.val:
                return False
            left = left.next
            right = right.next

        return True
